sycl-fusion/passes/syclcp/SYCLCP.cpp

```diff
@@ -0,0 +1,260 @@
+    // Advance the pointer
+    *ValPtr = *ValPtr + sizeof(float);
+    return ConstantFP::get(Ty, DoubleValue);
+  }
+  return createStringError(inconvertibleErrorCode(),
+                           "Only scalar and byval aggregate constants can be "
+                           "propagated by -sycl-cp");
+}
+
+///
+/// Initialize the members of an (potentially nested) aggregate constant through
+/// store instructions.
+static Error initializeAggregateConstant(const unsigned char **ValPtr,
+                                         Type *CurrentTy, IRBuilder<> &Builder,
+                                         Value *Alloca, Type *RootTy,
+                                         ArrayRef<Value *> Indices) {
+  if (CurrentTy->isIntegerTy() || CurrentTy->isFloatTy() ||
+      CurrentTy->isDoubleTy()) {
+    llvm::Expected<llvm::Value *> CVal =
+        getConstantValue(ValPtr, CurrentTy, false);
+    if (auto E = CVal.takeError()) {
+      return E;
+    }
+    auto *GEP = Builder.CreateInBoundsGEP(RootTy, Alloca, Indices);
+    Builder.CreateStore(CVal.get(), GEP);
+    return Error::success();
+  }
+  if (CurrentTy->isArrayTy()) {
+    ArrayType *ArrTy = cast<ArrayType>(CurrentTy);
+    for (uint32_t I = 0; I < ArrTy->getArrayNumElements(); ++I) {
+      SmallVector<Value *> ArrayIndices;
+      ArrayIndices.insert(ArrayIndices.begin(), Indices.begin(), Indices.end());
+      ArrayIndices.push_back(Builder.getInt32(I));
+      llvm::Error RetCode =
+          initializeAggregateConstant(ValPtr, ArrTy->getElementType(), Builder,
+                                      Alloca, RootTy, ArrayIndices);
+      if (RetCode) {
+        return RetCode;
+      }
+    }
+    return Error::success();
+  }
+  if (CurrentTy->isStructTy()) {
+    StructType *StructTy = cast<StructType>(CurrentTy);
+    SmallVector<Constant *> StructValues;
+    for (auto IdxElem : llvm::enumerate(StructTy->elements())) {
+      SmallVector<Value *> StructIndices;
+      StructIndices.insert(StructIndices.begin(), Indices.begin(),
+                           Indices.end());
+      StructIndices.push_back(Builder.getInt32(IdxElem.index()));
+      llvm::Error RetCode = initializeAggregateConstant(
+          ValPtr, IdxElem.value(), Builder, Alloca, RootTy, StructIndices);
+      if (RetCode) {
+        return RetCode;
+      }
+    }
+    return Error::success();
+  }
+  return createStringError(inconvertibleErrorCode(),
+                           "Cannot construct constant aggregate");
+}
+
+///
+/// Try to create an allocation and initialization via store instructions for a
+/// aggregate constant argument passed byval.
+static Expected<Value *> createAggregateConstant(const unsigned char **ValPtr,
+                                                 Type *Ty,
+                                                 IRBuilder<> &Builder) {
+  auto *Alloca = Builder.CreateAlloca(Ty);
+  SmallVector<Value *> Indices;
+  Indices.push_back(Builder.getInt32(0));
```

Comment:
`ArrayRef` is [constructible from a single element](https://llvm.org/doxygen/classllvm_1_1ArrayRef.html#a3b1f44186f9787d7ffacb54b62d6798c): we should use that instead of performing an allocation on heap for a single pointer.